File: 227.py
class Solution(object):
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        ans = 0
        op = '+'
        n = 0
        s = s + '+0'
        stack = []
        for ch in s:
            #print ans, op, n, ch
            if ch in ['+','-']:
                if stack:
                    n = self.calc(ans, op, n)
                    ans, op = stack.pop()
                ans = self.calc(ans, op, n)
                op = ch
                n = 0
            elif ch in ['*','/']:
                if not stack:
                    stack.append((ans, op))
                    ans = n
                else:
                    ans = self.calc(ans, op, n)
                op = ch
                n = 0
            elif ch >= '0' and ch <= '9':
                n *= 10
                n += int(ch)

        return ans


    def calc(self, ans, op, n):
        if not ans:
            ans = 0
        if ans == True:
            ans = 1
        if not n:
            n = 0
        if n == True:
            n = 1

        if op is '+':
            return int(ans + n)
        if op is '-':
            return int(ans - n)
        if op is '*':
            return int(ans * n)
        if op is '/':
            return int(ans / n)

        
if __name__ == '__main__':
    test = Solution()
    print test.calculate("3+2*2")# = 7
    print test.calculate("3/2 ")# = 1
    print test.calculate("3+5 / 2 ")# = 5
    print test.calculate("  30")
    print test.calculate("0")
    print test.calculate("0*0")
    print test.calculate("1 + 1")
    print test.calculate("1 + 1")

File: 227.java
import java.util.Stack;

public class Solution {
    public int calculate(String s) {

        int ans = 0;
        int n = 0;
        char op = '+';
        s += "+0";
        Stack stack = new Stack();
        for (char ch : s.toCharArray()) {
            switch (ch) {
                case '+':
                case '-':
                    if (!stack.isEmpty()) {
                        n = calc(ans, op, n);
                        ans = (int) stack.pop();
                        op = (char) stack.pop();
                    }
                    ans = calc(ans, op, n);
                    op = ch;
                    n = 0;
                    break;
                case '*':
                case '/':
                    if (!stack.isEmpty()) {
                        ans = calc(ans, op, n);
                    } else {
                        stack.push(op);
                        stack.push(ans);
                        ans = n;
                    }
                    op = ch;
                    n = 0;
                    break;
                case ' ':
                    break;
                default:
                    n *= 10;
                    n += ch - 0x30;
            }
        }
        return ans;
    }

    private int calc(int ans, char op, int n) {
        switch (op) {
            case '+':
                return ans + n;
            case '-':
                return ans - n;
            case '*':
                return ans * n;
            case '/':
                return ans / n;
        }
        return 0;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.calculate("0*0"));
        System.out.println(solution.calculate(" 1+1 "));
    }
}

